### PR TITLE
Add `f16` and `f128` const eval for casting operations

### PR DESCRIPTION
**Pull request by tgross35**
This is a continuation of https://github.com/rust-lang/rust/pull/126429 that should add missing operations.

### CODE REVIEW

**Comment by tgross35** [comment]
Still a WIP with the tests

@rustbot author
r? @RalfJung

File: src/tools/miri/tests/pass/float.rs [hunk 1]
```diff
@@ -151,6 +170,73 @@ fn basic() {
...
+    // assert_eq::<u16>(f16::NAN as u16, 0);
+    // assert_eq::<u16>((-f16::NAN) as u16, 0);
+
+    // f16 -> i32
+    assert_eq::<i32>(f16::MAX as i32, 65504);
+    assert_eq::<i32>(f16::MIN as i32, -65504);
+
+    // f16 -> u32
+    assert_eq::<u32>(f16::MAX as u32, 65504);
+    assert_eq::<u32>(f16::MIN as u32, 0);
```
Comment (RalfJung) [inline src/tools/miri/tests/pass/float.rs:236]: Does this match rustc tests, or is rustc not ready yet to run all of these?
I can't really check whether your values make sense; so far I always did this by just running the test with rustc and then making sure Miri behavior matches rustc behavior.

Comment (tgross35) [reply src/tools/miri/tests/pass/float.rs:236]: rustc should be able to run the `f16`/`f128` on at least Linux x86 and aarch64, but many other platforms are missing the f16/f128 symbols until compiler builtins gets updated. Not sure if you meant fully testing in CI or just locally.

However, I still don't have the constants needed to fully test casting on the new types. So I moved only the test refactoring to its own PR (https://github.com/rust-lang/rust/pull/126727), and am just going to close this one until https://github.com/rust-lang/rust/pull/126608 can land.

Comment (tgross35) [reply src/tools/miri/tests/pass/float.rs:236]: Also there are some ABI issues on x86 without SSE2, which might explain the failure at https://github.com/rust-lang/rust/actions/runs/9589520480/job/26443432727?pr=126718#step:25:3379 (haven't looked too much into it)

**Pull request closed without merging**